azurelinuxagent/ga/cgroupapi.py

```diff
@@ -342,11 +397,139 @@ def start_extension_command(self, extension_name, command, cmd_name, timeout, sh
+        Returns True if the provided controller is enabled at the provided cgroup.
+
+        There are two ways to determine if a controller is enabled at the provided cgroup:
+
+        1. For non-leaf cgroups, the cgroup.subtree_control shows space separated list of the controllers which are
+        enabled to control resource distribution from the cgroup to its children. All non-root "cgroup.subtree_control"
+        files can only contain controllers which are enabled in the parent's "cgroup.subtree_control" file.
+                $ cat /sys/fs/cgroup/cgroup.subtree_control
+                cpuset cpu io memory hugetlb pids rdma misc
+
+        2. For leaf cgroups, the cgroup.subtree_control file will be empty and the presence of "<controller>."
+        prefixed interface files at the path indicate the controller is enabled.
+                $ ls /sys/fs/cgroup/azure.slice/walinuxagent.service/
+                cgroup.controllers  cgroup.max.descendants  cgroup.threads  cpu.pressure    cpu.weight.nice      memory.high       memory.oom.group  memory.swap.current  memory.zswap.current  pids.peak
+                cgroup.events       cgroup.pressure         cgroup.type     cpu.stat        io.pressure          memory.low        memory.peak       memory.swap.events   memory.zswap.max
+                cgroup.freeze       cgroup.procs            cpu.idle        cpu.uclamp.max  memory.current       memory.max        memory.pressure   memory.swap.high     pids.current
+                cgroup.kill         cgroup.stat             cpu.max         cpu.uclamp.min  memory.events        memory.min        memory.reclaim    memory.swap.max      pids.events
+                cgroup.max.depth    cgroup.subtree_control  cpu.max.burst   cpu.weight      memory.events.local  memory.numa_stat  memory.stat       memory.swap.peak     pids.max
+
+        If either check is True, the controller is enabled at the cgroup. Check 1 is necessary because no controller
+        interface files exist at the root cgroup, even if the controller is enabled.
         """
-        return CGroupsApi._foreach_legacy_cgroup(lambda *_: None)
+        if cgroup_path is not None and controller is not None:
+            # Check that the controller is enabled in the cgroup.subtree_control file
+            enabled_controllers_file = os.path.join(cgroup_path, 'cgroup.subtree_control')
+            if os.path.exists(enabled_controllers_file):
+                enabled_controllers = fileutil.read_file(enabled_controllers_file).rstrip().split(" ")
+                if controller in enabled_controllers:
+                    return True
+
+            # Check that the controller interface files exist in the cgroup
+            if os.path.exists(cgroup_path):
+                for item in os.listdir(cgroup_path):
+                    if item.startswith(controller + '.'):
+                        return True
+
+        return False
+
+    def get_cgroup_mount_points(self):
```

Comment:
maybe rename to get_cgroup_controller_mount_points

In v2 api, get_cgroup_controller_mount_points just calls some get_cgroup_root_mount_point

Reply:
"mount point" was appropriate for v1, since the controllers are mounted independently of each other. This is basically the path of the root cgroup, maybe we should change to that?

get_root_cgroups, etc ?

Reply:
Also, the code for v1 is overcomplicated because it keeps track of the paths for each controller separately. This causes a lot of code repetition in the code, logs & telemetry, making the code and logs harder to read.

v2 has a simpler model. 

Instead of making v2 as complicated as v1, shouldn't we make v1 as simple as v2? Can you explore this option?

Initially the code was using the paths for each controller separately, since it was using the file system API directly and, for example, it needed to write the PID of the Agent to both the cpu and memory paths. Now we are using the systemd API and we work in terms of slices and scopes, which are 1 single entity (instead of 1 per controller). I think we use the actual filesystem paths only to collect the metrics.

It has been proposed that we should also monitor io and maybe other metrics. If we keep in the path of tracking each controller separately, this is going to get ever more complicated that it already is. v2 is a good chance to simplify this